cmd/kube-apiserver/app/server.go

```diff
@@ -583,7 +584,19 @@ func BuildStorageFactory(s *options.ServerRunOptions) (*serverstorage.DefaultSto
 		storageFactory.SetEtcdLocation(groupResource, servers)
 	}
 
-	if s.Etcd.EncryptionProviderConfigFilepath != "" {
+	// Set up encryption at rest via transformer overrides if a configuration file was provided.
+	if len(s.Etcd.EncryptionProviderConfigFilepath) != 0 {
+		cloud, err := cloudprovider.InitCloudProvider(s.CloudProvider.CloudProvider, s.CloudProvider.CloudConfigFile)
+		if err != nil {
+			return nil, err
+		}
+		// Returns an envelope Service implemented by a KMS supported by the cloud hosting k8s.
+		cloudKMSGetter := func(name string) (envelope.Service, error) {
+			return cloud.KeyManagementService(name)
+		}
+		// Register cloud in plugin registry in case a cloudprovided KMS provider is required.
+		encryptionconfig.KMSPluginRegistry.RegisterCloudProvidedKMSPlugin(cloudKMSGetter)
```

Comment:
mutating a global singleton here doesn't seem right. why wouldn't cloud providers register plugins for the names they handle via the normal `KMSPluginRegistry.Register("google-cloudkms", ...)` path?

Reply:
I brainstormed on this for some time. The cloud provider should ideally not use any methods outside `pkg/cloudprovider`, as I understand. To make the cloudprovider register its plugins into the plugin registry, it will need to talk to the registry.
The registry cannot be in `pkg/cloudprovider`. If you can suggest a place for a plugin registry where cloudproviders can access it, we can have a neat solution to this issue.